packages/theme/src/index.js

```diff
@@ -0,0 +1,170 @@
+// @flow
+import * as React from 'react'
+import { isBrowser } from '@emotion/utils'
+
+let canUseCSSVars =
+  isBrowser &&
+  window.CSS &&
+  window.CSS.supports &&
+  window.CSS.supports('--test', 0)
+
+type ThemeType =
+  | $ReadOnly<{
+      [key: string]: ThemeType
+    }>
+  | string
+  | $ReadOnlyArray<ThemeType>
+
+type Ret<Theme> = {
+  Consumer: React.ComponentType<{
+    children: Theme => React.Node
+  }>,
+  consume: <Props>(
+    render: (props: Props, theme: Theme, ref: React.Ref<*>) => React.Node
+  ) => React.ComponentType<Props>,
+  Provider: React.ComponentType<{
+    theme: Theme,
+    children: React.Node,
+    supportsCSSVariables?: boolean
+  }>,
+  Extender: React.ComponentType<{ children: React.Node }>
+}
+
+function getCSSVarUsageTheme(theme: ThemeType, currentPath: string) {
+  if (typeof theme === 'string') {
+    return `var(--theme-${currentPath})`
+  }
+  if (Array.isArray(theme)) {
+    return theme.map((val, i) =>
+      getCSSVarUsageTheme(val, currentPath + '-' + i)
+    )
+  }
+  let keys = Object.keys(theme)
+  let obj = {}
+  for (let i = 0; i < keys.length; i++) {
+    let key = keys[i]
+    obj[key] = getCSSVarUsageTheme(theme[key], currentPath + '-' + key)
+  }
+
+  return obj
+}
+
+function getInlineStyles(
+  theme: ThemeType,
+  currentPath: string,
+  stylesObj: { [string]: string }
+) {
+  if (typeof theme === 'string') {
+    stylesObj[`--theme-${currentPath}`] = theme
```

Comment:
Not really, I don't see why, I think using the paths directly would be helpful since people could see what value it's referencing.

Reply:
im worrying about potential clashes, those are global variables, we should at least prefix them with something unique for emotion (and possibly configurable per instance/emotion provider)

Reply:
Yep, agree about the prefix, hashing wouldn't help solve clashes though since I think one big problem could be having multiple of these themes and them clashing so it would probably be best to have it as an option in createTheme.